File: src/main/java/ru/org/linux/topic/TopicListController.java
/*
 * Copyright 1998-2015 Linux.org.ru
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package ru.org.linux.topic;

import org.joda.time.DateTime;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.servlet.ModelAndView;
import org.springframework.web.servlet.View;
import org.springframework.web.servlet.view.RedirectView;
import ru.org.linux.group.Group;
import ru.org.linux.group.GroupDao;
import ru.org.linux.group.GroupNotFoundException;
import ru.org.linux.section.Section;
import ru.org.linux.section.SectionNotFoundException;
import ru.org.linux.section.SectionService;
import ru.org.linux.site.ScriptErrorException;
import ru.org.linux.site.Template;
import ru.org.linux.user.UserErrorException;
import ru.org.linux.util.BadDateException;
import ru.org.linux.util.DateUtil;
import ru.org.linux.util.ServletParameterException;
import ru.org.linux.util.ServletParameterMissingException;

import javax.annotation.Nullable;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.net.URLEncoder;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

@Controller
public class TopicListController {
  @Autowired
  private SectionService sectionService;

  @Autowired
  private TopicListService topicListService;

  @Autowired
  private TopicPrepareService prepareService;

  // TODO: здесь должен быть сервис, а не DAO
  @Autowired
  private GroupDao groupDao;

  private ModelAndView mainTopicsFeedHandler(
    Section section,
    HttpServletRequest request,
    TopicListRequest topicListForm,
    HttpServletResponse response,
    @Nullable Group group
  ) throws Exception {
    checkRequestConditions(section, group, topicListForm);
    Template tmpl = Template.getTemplate(request);

    ModelAndView modelAndView = new ModelAndView("view-news");

    modelAndView.addObject("group", group);

    modelAndView.addObject("url", "view-news.jsp");
    modelAndView.addObject("section", section);

    modelAndView.addObject("archiveLink", section.getArchiveLink());

    setExpireHeaders(response, topicListForm.getYear(), topicListForm.getMonth());

    modelAndView.addObject("navtitle", calculateNavTitle(section, group, topicListForm));

    topicListForm.setOffset(
      topicListService.fixOffset(topicListForm.getOffset())
    );

    List<Topic> messages = topicListService.getTopicsFeed(
      section,
      group,
      null,
      topicListForm.getOffset(),
      topicListForm.getYear(),
      topicListForm.getMonth(),
      20
    );

    modelAndView.addObject(
      "messages",
      prepareService.prepareMessagesForUser(
              messages,
              request.isSecure(),
              tmpl.getCurrentUser(),
              tmpl.getProf(),
              false
      )
    );

    modelAndView.addObject("offsetNavigation", topicListForm.getMonth() == null);

    return modelAndView;
  }

  @RequestMapping("/gallery/")
  public ModelAndView gallery(
    HttpServletRequest request,
    TopicListRequest topicListForm,
    HttpServletResponse response
  ) throws Exception {
    Section section = sectionService.getSection(Section.SECTION_GALLERY);

    ModelAndView modelAndView = mainTopicsFeedHandler(section, request, topicListForm, response, null);

    modelAndView.addObject("ptitle", calculatePTitle(section, topicListForm));
    modelAndView.addObject("url", "/gallery/");
    modelAndView.addObject("rssLink", "section-rss.jsp?section=3");

    return modelAndView;
  }

  @RequestMapping("/forum/lenta")
  public ModelAndView forum(
    HttpServletRequest request,
    TopicListRequest topicListForm,
    HttpServletResponse response
  ) throws Exception {
    Section section = sectionService.getSection(Section.SECTION_FORUM);

    ModelAndView modelAndView = mainTopicsFeedHandler(section, request, topicListForm, response, null);
    modelAndView.addObject("ptitle", calculatePTitle(section, topicListForm));
    modelAndView.addObject("url", "/forum/lenta");
    modelAndView.addObject("rssLink", "section-rss.jsp?section=2");

    return modelAndView;
  }

  @RequestMapping("/polls/")
  public ModelAndView polls(
    HttpServletRequest request,
    TopicListRequest topicListForm,
    HttpServletResponse response
  ) throws Exception {
    Section section = sectionService.getSection(Section.SECTION_POLLS);
    ModelAndView modelAndView = mainTopicsFeedHandler(section, request, topicListForm, response, null);

    modelAndView.addObject("url", "/polls/");
    modelAndView.addObject("params", null);
    modelAndView.addObject("ptitle", calculatePTitle(section, topicListForm));

    return modelAndView;
  }

  @RequestMapping("/news/")
  public ModelAndView news(
    HttpServletRequest request,
    TopicListRequest topicListForm,
    HttpServletResponse response
  ) throws Exception {
    Section section = sectionService.getSection(Section.SECTION_NEWS);
    ModelAndView modelAndView = mainTopicsFeedHandler(section, request, topicListForm, response, null);

    modelAndView.addObject("url", "/news/");
    modelAndView.addObject("ptitle", calculatePTitle(section, topicListForm));
    modelAndView.addObject("rssLink", "section-rss.jsp?section=1");

    return modelAndView;
  }

  @RequestMapping("/gallery/{group}")
  public ModelAndView galleryGroup(
    HttpServletRequest request,
    TopicListRequest topicListForm,
    @PathVariable("group") String groupName,
    HttpServletResponse response
  ) throws Exception {
    Section section = sectionService.getSection(Section.SECTION_GALLERY);

    return group(section, request, topicListForm, groupName, response);
  }

  @RequestMapping("/news/{group}")
  public ModelAndView newsGroup(
    HttpServletRequest request,
    TopicListRequest topicListForm,
    @PathVariable("group") String groupName,
    HttpServletResponse response
  ) throws Exception {
    Section section = sectionService.getSection(Section.SECTION_NEWS);

    return group(section, request, topicListForm, groupName, response);
  }

  @RequestMapping("/polls/{group}")
  public ModelAndView pollsGroup(
    HttpServletRequest request,
    TopicListRequest topicListForm,
    @PathVariable("group") String groupName,
    HttpServletResponse response
  ) throws Exception {
    Section section = sectionService.getSection(Section.SECTION_POLLS);

    return group(section, request, topicListForm, groupName, response);
  }

  @RequestMapping("/{section}/archive/{year}/{month}")
  public ModelAndView galleryArchive(
    HttpServletRequest request,
    @PathVariable String section,
    @PathVariable int year,
    @PathVariable int month,
    HttpServletResponse response
  ) throws Exception {
    TopicListRequest topicListForm = new TopicListRequest();

    Section sectionObject = sectionService.getSectionByName(section);

    if (sectionObject.isPremoderated()) {
      topicListForm.setYear(year);
      topicListForm.setMonth(month);

      ModelAndView modelAndView = mainTopicsFeedHandler(sectionObject, request, topicListForm, response, null);

      modelAndView.addObject("ptitle", calculatePTitle(sectionObject, topicListForm));

      return modelAndView;
    } else {
      return new ModelAndView(new RedirectView(sectionObject.getSectionLink()));
    }
  }

  @RequestMapping(value = "/show-topics.jsp", method = RequestMethod.GET)
  public View showUserTopics(
    @RequestParam("nick") String nick,
    @RequestParam(value = "output", required = false) String output
  ) {
    if (output != null) {
      return new RedirectView("/people/" + nick + "/?output=rss");
    }

    return new RedirectView("/people/" + nick + '/');
  }

  @Deprecated
  @RequestMapping(value = "/view-news.jsp", params = {"section", "!tag"})
  public View oldLink(
    TopicListRequest topicListForm,
    @RequestParam("section") int sectionId,
    @RequestParam(value="group", defaultValue = "0") int groupId
  ) throws Exception {
    Section section = sectionService.getSection(sectionId);

    StringBuilder redirectLink = new StringBuilder();

    redirectLink.append(section.getNewsViewerLink());

    if (topicListForm.getYear() != null && topicListForm.getMonth() != null) {
      redirectLink
        .append(topicListForm.getYear())
        .append('/')
        .append(topicListForm.getMonth());
    } else if (groupId>0) {
      Group group = groupDao.getGroup(groupId);
      redirectLink
        .append(group.getUrlName())
        .append('/');
    }

    String queryStr = topicListForm.getOffset() == null ?
        "" :
        URLEncoder.encode("offset=" + topicListForm.getOffset(), "UTF-8");

    if (!queryStr.isEmpty()) {
      redirectLink
        .append('?')
        .append(queryStr);
    }

    return new RedirectView(redirectLink.toString());
  }

  @RequestMapping("/section-rss.jsp")
  public ModelAndView showRSS(
    HttpServletRequest request,
    TopicListRequest topicListForm,
    @RequestParam(value="section", defaultValue = "1") int sectionId,
    @RequestParam(value="group", defaultValue = "0") int groupId,
    @RequestParam(value="filter", required = false) String filter
  ) throws Exception {

    final String[] filterValues = {"all", "notalks", "tech"};
    final Set<String> filterValuesSet = new HashSet<>(Arrays.asList(filterValues));

    if (filter != null && !filterValuesSet.contains(filter)) {
      throw new UserErrorException("Некорректное значение filter");
    }
    boolean notalks = filter != null && "notalks".equals(filter);
    boolean tech = filter != null && "tech".equals(filter);

    Section section = sectionService.getSection(sectionId);
    String ptitle = section.getName();

    Group group = null;
    if (groupId != 0) {
      group = groupDao.getGroup(groupId);
      ptitle += " - " + group.getTitle();
    }

    checkRequestConditions(section, group, topicListForm);

    ModelAndView modelAndView = new ModelAndView("section-rss");

    modelAndView.addObject("group", group);
    modelAndView.addObject("section", section);
    modelAndView.addObject("ptitle", ptitle);

    DateTime fromDate = DateTime.now().minusMonths(3);

    List<Topic> messages = topicListService.getRssTopicsFeed(section, group, fromDate.toDate(), notalks, tech);

    modelAndView.addObject("messages", prepareService.prepareMessages(messages, request.isSecure()));
    return modelAndView;
  }

  private ModelAndView group(
    Section section,
    HttpServletRequest request,
    TopicListRequest topicListForm,
    String groupName,
    HttpServletResponse response
  ) throws Exception {
    Group group = groupDao.getGroup(section, groupName);

    ModelAndView modelAndView = mainTopicsFeedHandler(
      section,
      request,
      topicListForm,
      response,
      group
    );

    modelAndView.addObject("ptitle", section.getName() + " - " + group.getTitle());

    modelAndView.addObject("url", group.getUrl());

    return modelAndView;
  }

  public static void setExpireHeaders(
          HttpServletResponse response,
          Integer year, Integer month
  ) {
    if (month == null) {
      response.setDateHeader("Expires", System.currentTimeMillis() + 60 * 1000);
      response.setDateHeader("Last-Modified", System.currentTimeMillis());
    } else {
      DateTime endOfMonth = new DateTime(year, month, 1, 0, 0).plusMonths(1);

      if (endOfMonth.isBeforeNow()) {
        response.setDateHeader("Last-Modified", endOfMonth.getMillis());
      } else {
        response.setDateHeader("Expires", System.currentTimeMillis() + 60 * 1000);
        response.setDateHeader("Last-Modified", System.currentTimeMillis());
      }
    }
  }

  private void checkRequestConditions(Section section, Group group, TopicListRequest topicListForm)
    throws Exception {
    if (topicListForm.getMonth() != null && topicListForm.getYear() == null) {
      throw new ServletParameterMissingException("year");
    }
    if (section == null) {
      throw new ServletParameterException("section or tag required");
    }
    if (group != null && group.getSectionId() != section.getId()) {
      throw new ScriptErrorException("группа #" + group.getId() + " не принадлежит разделу #" + section.getId());
    }
  }

  private static String calculatePTitle(Section section, TopicListRequest topicListForm)
    throws BadDateException {

    if (topicListForm.getMonth() == null) {
      return section.getName();
    } else {
      return "Архив: " + section.getName()
              + ", " +
              topicListForm.getYear() +
              ", " +
              DateUtil.getMonth(topicListForm.getMonth());
    }
  }

  private static String calculateNavTitle(Section section, Group group, TopicListRequest topicListForm)
    throws BadDateException, SectionNotFoundException {

    StringBuilder navTitle = new StringBuilder();

    if (group == null) {
      navTitle.setLength(0);
      navTitle.append(section.getName());
    } else {
      navTitle.setLength(0);
      navTitle
              .append(section.getName())
              .append(" «")
              .append(group.getTitle())
              .append("»");
    }

    if (topicListForm.getMonth() != null) {
      navTitle
        .append(" - Архив ")
        .append(topicListForm.getYear())
        .append(", ")
        .append(DateUtil.getMonth(topicListForm.getMonth()));
    }
    return navTitle.toString();
  }

  @ExceptionHandler({GroupNotFoundException.class, SectionNotFoundException.class})
  @ResponseStatus(HttpStatus.NOT_FOUND)
  public ModelAndView handleNotFoundException() {
    return new ModelAndView("errors/code404");
  }
}

File: src/main/scala/ru/org/linux/topic/TopicListController.scala
/*
 * Copyright 1998-2023 Linux.org.ru
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package ru.org.linux.topic

import com.typesafe.scalalogging.StrictLogging
import org.joda.time.DateTime
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.*
import org.springframework.web.context.request.WebRequest
import org.springframework.web.servlet.view.RedirectView
import org.springframework.web.servlet.{ModelAndView, View}
import ru.org.linux.auth.AuthUtil.AuthorizedOpt
import ru.org.linux.group.{Group, GroupDao, GroupNotFoundException, GroupPermissionService}
import ru.org.linux.section.{Section, SectionController, SectionNotFoundException, SectionService}
import ru.org.linux.site.{ScriptErrorException, Template}
import ru.org.linux.tag.{TagPageController, TagService}
import ru.org.linux.topic.TopicListController.ForumFilter.{NoTalks, Tech}
import ru.org.linux.topic.TopicListController.{ForumFilter, ForumFilters, calculatePTitle}
import ru.org.linux.user.UserErrorException
import ru.org.linux.util.{DateUtil, ServletParameterException}

import java.util.concurrent.CompletionStage
import javax.annotation.Nullable
import scala.compat.java8.FutureConverters.*
import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future
import scala.jdk.CollectionConverters.*

object TopicListController {
  private def calculatePTitle(section: Section, groupOpt: Option[Group], topicListForm: TopicListRequest): String = {
    groupOpt match {
      case Some(group) =>
        s"${section.getName} - ${group.title}"
      case None =>
        if (topicListForm.yearMonth.isEmpty) {
          section.getName + topicListForm.filter.map(f => s" (${f.title})").getOrElse("")
        } else {
          s"Архив: ${section.getName}, ${topicListForm.getYear.get}, ${DateUtil.getMonth(topicListForm.getMonth.get)}"
        }
    }
  }

  private def calculateNavTitle(section: Section, group: Option[Group], topicListForm: TopicListRequest): String = {
    val navTitle = new StringBuilder(section.getName)

    group.foreach { group =>
      navTitle.append(s" «${group.title}»")
    }

    topicListForm.filter.foreach { f =>
      navTitle.append(s" (${f.title})")
    }

    if (topicListForm.getMonth.isDefined) {
      navTitle.append(" - Архив ").append(topicListForm.getYear.get).append(", ").append(DateUtil.getMonth(topicListForm.getMonth.get))
    }

    navTitle.toString
  }

  sealed trait ForumFilter {
    def id: String
    def title: String

    final def getId: String = id
    final def getTitle: String = title
  }

  object ForumFilter {
    case object NoTalks extends ForumFilter {
      override val id = "notalks"
      override val title = "без talks"
    }

    case object Tech extends ForumFilter {
      override val id = "tech"
      override val title = "тех. форум"
    }
  }

  private val ForumFilters = Seq(NoTalks, Tech)
}

@Controller
class TopicListController(sectionService: SectionService, topicListService: TopicListService,
                          prepareService: TopicPrepareService, tagService: TagService,
                          groupDao: GroupDao, groupPermissionService: GroupPermissionService) extends StrictLogging {
  private def mainTopicsFeedHandler(section: Section, topicListForm: TopicListRequest,
                                    group: Option[Group]): Future[ModelAndView] = AuthorizedOpt { currentUserOpt =>
    val deadline = TagPageController.Timeout.fromNow

    checkRequestConditions(section, group)

    val activeTagsF = if (topicListForm.yearMonth.isEmpty) {
      tagService.getActiveTopTags(section, group, topicListForm.filter, deadline)
    } else {
      Future.successful(Seq.empty)
    }

    val modelAndView = new ModelAndView("view-news")

    modelAndView.addObject("ptitle", calculatePTitle(section, group, topicListForm))

    modelAndView.addObject("topicListRequest", topicListForm)

    group foreach { group =>
      modelAndView.addObject("group", group)
    }

    modelAndView.addObject("section", section)
    modelAndView.addObject("archiveLink", section.getArchiveLink)

    if (section.getId != Section.SECTION_FORUM) {
      modelAndView.addObject("groupList",
        SectionController.groupsSorted(groupDao.getGroups(section).asScala).asJava)
    } else {
      modelAndView.addObject("filters", ForumFilters.asJava)
      modelAndView.addObject("filter", topicListForm.filter.getOrElse(""))
    }

    modelAndView.addObject("navtitle", TopicListController.calculateNavTitle(section, group, topicListForm))

    val tmpl = Template.getTemplate

    val messages = topicListService.getTopicsFeed(section, group, None, topicListForm.offset,
      topicListForm.yearMonth, 20, currentUserOpt.map(_.user), topicListForm.filter.contains(NoTalks),
      topicListForm.filter.contains(Tech))

    modelAndView.addObject(
      "messages",
      prepareService.prepareTopicsForUser(messages, currentUserOpt.map(_.user), tmpl.getProf, loadUserpics = false))

    modelAndView.addObject("offsetNavigation", topicListForm.yearMonth.isEmpty)

    val addUrl = group match {
      case Some(group) if groupPermissionService.isTopicPostingAllowed(group, currentUserOpt.map(_.user).orNull) =>
        AddTopicController.getAddUrl(group)
      case None if groupPermissionService.isTopicPostingAllowed(section, currentUserOpt.map(_.user)) =>
        AddTopicController.getAddUrl(section)
      case _ =>
        ""
    }

    modelAndView.addObject("addUrl", addUrl)

    activeTagsF.map { activeTags =>
      if (activeTags.nonEmpty) {
        modelAndView.addObject("activeTags", activeTags.asJava)
      }

      modelAndView
    }
  }

  @RequestMapping(path = Array("/{section:(?:news)|(?:polls)|(?:articles)|(?:gallery)}/"))
  def topics(@PathVariable("section") sectionName: String,
             @RequestParam(value="offset", defaultValue = "0") offset: Int): CompletionStage[ModelAndView] = {
    val section = sectionService.getSectionByName(sectionName)

    val topicListForm = TopicListRequest.ofOffset(offset)

    mainTopicsFeedHandler(section, topicListForm, None).map { modelAndView =>
      modelAndView.addObject("url", section.getNewsViewerLink)
      modelAndView.addObject("rssLink", s"section-rss.jsp?section=${section.getId}")
    }.toJava
  }

  private def parseFilter(@Nullable value: String): Option[ForumFilter] = {
    Option(value)
      .map(v => ForumFilters.find(_.id == v).getOrElse(throw new UserErrorException("Некорректное значение filter")))
  }

  @RequestMapping(path = Array("/forum/lenta"))
  def forum(@RequestParam(value="offset", defaultValue = "0") offset: Int,
            @RequestParam(value = "filter", required = false) filter: String): CompletionStage[ModelAndView] = {
    val section = sectionService.getSection(Section.SECTION_FORUM)

    val topicListForm = TopicListRequest.ofOffset(offset).copy(filter = parseFilter(filter))

    mainTopicsFeedHandler(section, topicListForm, None).map { modelAndView =>
      if (filter==null) {
        modelAndView.addObject("url", section.getNewsViewerLink)
        modelAndView.addObject("rssLink", s"section-rss.jsp?section=${section.getId}")
      } else {
        modelAndView.addObject("url", section.getNewsViewerLink + s"?filter=$filter")
        modelAndView.addObject("rssLink", s"section-rss.jsp?section=${section.getId}&filter=$filter")
      }
    }.toJava
  }

  @RequestMapping(path = Array("/{section:(?:news)|(?:polls)|(?:articles)|(?:gallery)}/{group:[^.]+}"))
  def topicsByGroup(@PathVariable("section") sectionName: String,
                    @RequestParam(value="offset", defaultValue = "0") offset: Int,
                    @PathVariable("group") groupName: String): CompletionStage[ModelAndView] = {
    val section = sectionService.getSectionByName(sectionName)

    val group = groupDao.getGroup(section, groupName)

    val topicListForm = TopicListRequest.ofOffset(offset)

    mainTopicsFeedHandler(section, topicListForm, Some(group)).map { modelAndView =>
      modelAndView.addObject("url", group.getUrl)
    }.toJava
  }

  @RequestMapping(path = Array("/{section}/archive/{year:\\d{4}}/{month}"))
  def sectionArchive(@PathVariable section: String, @PathVariable year: Int, @PathVariable month: Int): CompletionStage[ModelAndView] = {
    val sectionObject = sectionService.getSectionByName(section)

    (if (sectionObject.isPremoderated) {
      val topicListForm = TopicListRequest.orYearMonth(year, month)

      mainTopicsFeedHandler(sectionObject, topicListForm, None)
    } else {
      Future.successful(new ModelAndView(new RedirectView(sectionObject.getSectionLink)))
    }).toJava
  }

  @RequestMapping(value = Array("/show-topics.jsp"), method = Array(RequestMethod.GET))
  def showUserTopics(@RequestParam("nick") nick: String,
                     @RequestParam(value = "output", required = false) output: String): View = {
    if (output != null) {
      new RedirectView(s"/people/$nick/?output=rss")
    } else {
      new RedirectView(s"/people/$nick/")
    }
  }

  @RequestMapping(path = Array("/section-rss.jsp"))
  def showRSS(@RequestParam(value = "section", defaultValue = "1") sectionId: Int,
              @RequestParam(value = "group", defaultValue = "0") groupId: Int,
              @RequestParam(value = "filter", required = false) filter: String,
              webRequest: WebRequest): ModelAndView = {
    val forumFilter = parseFilter(filter)

    val section = sectionService.getSection(sectionId)
    var ptitle = section.getName

    val group = if (groupId != 0) {
      val g = groupDao.getGroup(groupId)
      ptitle += " - " + g.title
      Some(g)
    } else {
      None
    }

    forumFilter.foreach { f =>
      ptitle += s" (${f.title})"
    }

    checkRequestConditions(section, group)
    val modelAndView = new ModelAndView("section-rss")

    group.foreach { group =>
      modelAndView.addObject("group", group)
    }

    modelAndView.addObject("section", section)
    modelAndView.addObject("ptitle", ptitle)

    val notalks = forumFilter.contains(NoTalks)
    val tech = forumFilter.contains(Tech)

    val fromDate = DateTime.now.minusMonths(3)
    val messages = topicListService.getRssTopicsFeed(section, group, fromDate.toDate, notalks, tech)

    // не лучший вариант, так как включает комментарии
    // по хорошему тут надо учитывать только правки текста топика
    val lastModified = messages.view.map(_.lastModified.getTime).maxOption

    if (lastModified.exists(webRequest.checkNotModified)) {
      null
    } else {
      modelAndView.addObject("messages", prepareService.prepareTopics(messages.toSeq).asJava)

      modelAndView
    }
  }

  private def checkRequestConditions(section: Section, group: Option[Group]): Unit = {
    if (section == null) {
      throw new ServletParameterException("section or tag required")
    }

    group foreach { group =>
      if (group.sectionId != section.getId) {
        throw new ScriptErrorException(s"группа #${group.id} не принадлежит разделу #${section.getId}")
      }
    }
  }

  @ExceptionHandler(Array(classOf[GroupNotFoundException], classOf[SectionNotFoundException]))
  @ResponseStatus(HttpStatus.NOT_FOUND)
  def handleNotFoundException: ModelAndView = new ModelAndView("errors/code404")
}